app/javascript/src/Policies/components/CustomPolicy.jsx

```diff
@@ -0,0 +1,155 @@
+  const props = {
+    name: getMetaContent('csrf-param'),
+    value: getMetaContent('csrf-token'),
+    type: 'hidden'
+  }
+  return (
+    <input {...props} />
+  )
+}
+
+function Editor ({onChange, code}: {onChange: (string) => void, code: Object}): React.Node {
+  const [ state, setState ] = useState({valid: true, code: toJson(code)})
+
+  const onCodeChange = (editor, metadata, code) => {
+    setState({ valid: true, code })
+    try {
+      onChange(fromJson(code))
+    } catch (err) {
+      setState({ valid: false, code })
+    }
+  }
+
+  const [ icon, cls, hiddenClass, errorClass ] = state.valid
+    ? [ 'check', 'valid', 'is-hidden', '' ]
+    : [ 'times', 'invalid', '', 'CustomPolicy-error' ]
+
+  return (
+    <div className={`${errorClass} panel panel-default`}>
+      <div className="panel-heading">
+        <i className={`${cls} fa fa-${icon}`} />
+        {' JSON Schema'}
+      </div>
+      <CodeMirror
+        value={state.code}
+        onChange={onCodeChange}
+        autoCursor={false}
+        options={CM_OPTIONS}
+      />
+      <div className={hiddenClass}> There's an error in the JSON Schema</div>
+    </div>
+  )
+}
+
+function FormInput (
+  props: {
+    type: 'text' | 'textarea',
+    humanname: string,
+    name: string,
+    value?: string,
+    onChange: OnChange,
+    disabled?: boolean
+  }) {
+  return (
+    <label>
+      {`${props.humanname}:`}
+      {(props.type === 'text') ? <input type="text" {...props} /> : <textarea {...props} /> }
+    </label>
+  )
+}
+
+function CustomPolicyForm ({policy, onChange, isNewPolicy}: {policy: RegistryPolicy, onChange: OnChange, isNewPolicy: boolean}): React.Node {
+  const action = (isNewPolicy) ? '/p/admin/registry/policies/' : `/p/admin/registry/policies/${policy.name}-${policy.version}`
+  return (
+    <form action={action} method="post">
+      <FormInput type="text" name="name" humanname="Name" value={policy.name} onChange={onChange} disabled={!isNewPolicy} />
+      <FormInput type="text" name="humanName" humanname="Human Name" value={policy.humanName} onChange={onChange} />
+      <FormInput type="text" name="version" humanname="Version" value={policy.version} onChange={onChange} disabled={!isNewPolicy} />
+      <FormInput type="textarea" name="summary" humanname="Summary" value={policy.summary} onChange={onChange} />
+      <FormInput type="textarea" name="description" humanname="Description" value={policy.description} onChange={onChange} />
+      <input name="configuration" type="hidden" value={JSON.stringify(policy.configuration)} />
+      {(!isNewPolicy) ? <input name="_method" type="hidden" value='put' /> : ''}
+      <input type="submit" />
+      <CSRFToken />
+    </form>
+  )
+}
+
+function Form ({policy}: {policy: RegistryPolicy}): React.Node {
+  const [pol, setPolicy] = useState(policy)
+  const onSchemaEdited = (pol: RegistryPolicy) => (configuration: string) => setPolicy({...pol, ...{configuration}})
```

Comment:
I think this doesn't need to be a double function, `pol` comes from the state anyway. Would not it be an equivalent to `this.state.pol`?